### PR TITLE
Avoid lost messages by acknowledges message receipt after the message…

### PR DESCRIPTION
**Pull request by charlesmchen**
… is processed.

Another change that will require careful testing.

PTAL @michaelkirk

### CODE REVIEW

File: src/Messages/TSMessagesManager.h [hunk 1]
```diff
@@ -27,18 +29,8 @@ NS_ASSUME_NONNULL_BEGIN
...
-- (void)handleReceivedEnvelope:(OWSSignalServiceProtosEnvelope *)envelope;
-
-/**
- * Processes all kinds of incoming envelopes with a data message, along with any attachments.
- *
- * @returns
- *   If an incoming message is created, it will be returned. If it is, for example, a group update,
- *   no incoming message is created, so nil will be returned.
- */
-- (TSIncomingMessage *)handleReceivedEnvelope:(OWSSignalServiceProtosEnvelope *)envelope
```
Comment (charlesmchen) [inline src/Messages/TSMessagesManager.h:39]: This method was only used internally, so I've removed it from the header.

File: src/Messages/TSMessagesManager.m [hunk 2]
```diff
@@ -241,14 +259,17 @@ - (void)handleSecureMessageAsync:(OWSSignalServiceProtosEnvelope *)messageEnvelo
                         [TSErrorMessage missingSessionWithEnvelope:messageEnvelope withTransaction:transaction];
                     [errorMessage saveWithTransaction:transaction];
                 }];
+                DDLogError(@"Skipping message envelope for unknown session.");
+                completion(nil);
```
Comment (charlesmchen) [inline src/Messages/TSMessagesManager.m:263]: There were lots of places in this method were we never called the completion handler.

File: src/Messages/TSMessagesManager.m [hunk 3]
```diff
@@ -297,6 +318,7 @@ - (void)handlePreKeyBundleAsync:(OWSSignalServiceProtosEnvelope *)preKeyEnvelope
         NSData *encryptedData = preKeyEnvelope.hasContent ? preKeyEnvelope.content : preKeyEnvelope.legacyMessage;
         if (!encryptedData) {
             DDLogError(@"Skipping message envelope which had no encrypted data");
+            completion(nil);
```
Comment (charlesmchen) [inline src/Messages/TSMessagesManager.m:321]: Same for this method.

**Comment by charlesmchen** [comment]
See: https://github.com/WhisperSystems/Signal-iOS/pull/1985

File: src/Network/WebSockets/TSSocketManager.m [hunk 4]
```diff
@@ -381,15 +379,22 @@ - (void)processWebSocketRequestMessage:(WebSocketRequestMessage *)message {
 
         if (!decryptedPayload) {
             DDLogWarn(@"Failed to decrypt incoming payload or bad HMAC");
```
Comment (michaelkirk) [inline src/Network/WebSockets/TSSocketManager.m:381]: I think we should display a user facing error here, it's arguably out of scope, so feel free to add a new card.

**Review by michaelkirk — APPROVED** [review]
LGTM

File: src/Network/WebSockets/TSSocketManager.m (hunk 4 above)
Comment (charlesmchen) [reply src/Network/WebSockets/TSSocketManager.m:381]: Added a card.

**Pull request merged**